### ISSUE TITLE
[Feature request] Make results of initial scan available

### ISSUE THREAD

**Issue by Shelim** [notify-rs/notify#478]
This is usecase for #339 

Scenario: 
 - Platform with slow file-system access and a lot of files
 - Need initial list of files as well as events on changes

Now I cannot obtain initial list of files from the API. I have to call separately `walkdir` which imply some drawbacks:

1. I need to rescan the whole file-system *again*
2. If one file is changed during rescanning, it changes can be lost (initial scan output will be provided later then changed event)
3. It imply memory cost for rescanning large file-system again
4. There may be some platform-specific inconsistencies in case symlink/hardlink. Initial set of files from notify may differ from `walkdir` output (unless notify is using `walkdir` internally, I haven't check up that)

**Comment by 0xpr03** [comment]
There is no initial scan on windows. So this would be a special thing for inotify / pollwatcher.

**Comment by Shelim** [comment]
Ah, I tricked myself the notify is doing initial scan as I only used pollwatcher. Okay, it seems legit now

**Comment by ShayBox** [comment]
This would be useful, I already use walkdir prior to notify, this would prevent double scanning, speeding up my program startup, and reduce dependencies in my project.

**Comment by 0xpr03** [comment]
How exactly would you like to access the initial scan ? Or do you want to access the live state too ?

One could want access to the full internal path cache, or alternatively provide a callback function that gets called for each path found during the initial scan.

**Comment by ShayBox** [comment]
I suppose whichever is easier, pollwatcher uses walkdir internally so exposing that would probably be easiest, but I'm not sure

**Comment by Shelim** [comment]
If we can change emitted event kinds, I believe the best way would be to add event callback kind for `initial` state. This way you can use exactly the same implementation mechanism you use for updated files

**Comment by 0xpr03** [comment]
Could you try out #507 and give some feedback ? There is an example inside too.

**Comment by ShayBox** [comment]
Is there a way to optionally supply the same handler for both?
```rust
let mut watcher = PollWatcher::with_initial_scan(tx, notify_config, Some(tx))?;
```

**Comment by 0xpr03** [comment]
Hm yeah, I opted for putting the Result inside the ScanEvent type - maybe that's more confusing than anything (changing this for notify would be a semver break).

**Comment by 0xpr03** [comment]
I left it like that and added some info to the ScanEventHandler docs.

**Comment by ShayBox** [comment]
Returning a result is fine, I just meant that normal event is wrapped in a result, while a scan event includes the result not just the pathbuf

Whats the purpose of the option parameter since the with variant is optional to use?

**Comment by 0xpr03** [comment]
> Whats the purpose of the option parameter since the with variant is optional to use?

[Code unification](https://github.com/notify-rs/notify/pull/507/files#diff-432d682b91f875b61c8b78650fa9bd8cd21cd9c2dd510479c5d8a490a961bccbR496)

> while a scan event includes the result not just the pathbuf

Yeah changing this to use the result in the normal event would be a major version change. But for the scan events it definitely makes sense to have this as one type.

**Comment by 0xpr03** [comment]
Hm I could internalize this so you don't have to add a `Some`, why not.

**Comment by Shelim** [comment]
Yeah, since we already include `with_initial_scan` in the function name :-)

I looked at the example and it seems to be exactly what I needed. I will check the implementation in actual app in a few moments

**Comment by ShayBox** [comment]
This is what I came up with but `'Sender<PathBuf>' cannot be shared between threads safely`

```rust
use std::{
    path::{Path, PathBuf},
    time::Duration,
};

use notify::{Config, Event, PollWatcher, RecursiveMode, Watcher};

fn main() {
    let (tx, rx) = std::sync::mpsc::channel();
    let mut watcher = PollWatcher::with_initial_scan(
        |event_result: notify::Result<Event>| {
            let Ok(event) = event_result else {
                return;
            };

            for path in event.paths {
                tx.send(path).unwrap();
            }
        },
        Config::default().with_poll_interval(Duration::from_secs(1)),
        Some(|path_result: notify::Result<PathBuf>| {
            let Ok(path) = path_result else {
                return;
            };

            tx.send(path).unwrap();
        }),
    )
    .unwrap();

    watcher
        .watch(Path::new("."), RecursiveMode::Recursive)
        .unwrap();

    loop {
        let path = rx.recv().unwrap();
        println!("path: {:?}", path);
    }
}
```

**Comment by 0xpr03** [comment]
@ShayBox 
```rust
fn main() {
    let (tx, rx) = std::sync::mpsc::channel();
    let tx_c = tx.clone();
    let mut watcher = PollWatcher::with_initial_scan(
        move |event_result: notify::Result<Event>| {
            let Ok(event) = event_result else {
                return;
            };

            for path in event.paths {
                tx_c.send(path).unwrap();
            }
        },
        Config::default().with_poll_interval(Duration::from_secs(1)),
        move |path_result: notify::Result<PathBuf>| {
            let Ok(path) = path_result else {
                return;
            };

            tx.send(path).unwrap();
        },
    )
    .unwrap();

    watcher
        .watch(Path::new("."), RecursiveMode::Recursive)
        .unwrap();

    loop {
        let path = rx.recv().unwrap();
        println!("path: {:?}", path);
    }
}
```
See also the [example](https://github.com/notify-rs/notify/pull/507/files#diff-8315e29caa00e27d8b965c2bab2ef4f4ab6d7bff2598620a579e0740d1cf6769R31-R42). You need to `move` and clone() your tx once, as you need one per closure.

**Comment by ShayBox** [comment]
:+1: It works for me, I can remove walkdir, rayon, and a few functions from my project now

**Comment by Shelim** [comment]
Looks good indeed, thank you!

**Comment by ShayBox** [comment]
Is there a planned release soon or can a release be made with these changes? I want to publish to crates.io but I can't use git dependencies on crates.io

**Comment by 0xpr03** [comment]
I have a release planned, there just some PRs I want to merge first
